### PR TITLE
Replace `TickType` with customizable `TickPosition`

### PR DESCRIPTION
**Pull request by Gowsky**
This PR solves  #106.

There are two `TickPosition`s:
1. `Edge`, which replaces `TickType.Minor`. It works exactly the same as before.
2. `Center`, which replaces `TickType.Major`. Now it may have custom `offset` and `spacing`.

The column chart in the sample app now uses `TickPosition.Center` with `offset` equal to `1` and `spacing` equal to `3`.

### CODE REVIEW

File: vico/core/src/main/java/com/patrykandpatryk/vico/core/axis/horizontal/HorizontalAxis.kt [hunk 1]
```diff
@@ -242,9 +281,115 @@ public class HorizontalAxis<Position : AxisPosition.Horizontal>(
...
+            public constructor(spacing: Int) : this(offset = spacing, spacing = spacing)
+
+            init {
+                if (offset < 0) {
+                    throw IllegalArgumentException("The offset cannot be negative. Received $offset.")
+                }
+
+                if (spacing < 1) {
+                    throw IllegalArgumentException("The offset cannot be less than 1. Received $spacing.")
+                }
```
Comment (patrykmichalik) [inline vico/core/src/main/java/com/patrykandpatryk/vico/core/axis/horizontal/HorizontalAxis.kt:373]: I think we could use `require` here.

**Review by patrykmichalik — CHANGES_REQUESTED** [review]

File: vico/core/src/main/java/com/patrykandpatryk/vico/core/axis/horizontal/HorizontalAxis.kt [hunk 2]
```diff
@@ -60,50 +72,66 @@ public class HorizontalAxis<Position : AxisPosition.Horizontal>(
...
         var valueIndex: Float = chartValues.minX + scrollAdjustment * step
 
         for (index in 0 until tickCount) {
-            guideline?.takeIf {
-                it.fitsInVertical(
+            val shouldDraw = valueIndex >= tickPosition.offset &&
+                (valueIndex - tickPosition.offset) % tickPosition.spacing == 0f
+
+            guideline
+                ?.takeIf {
```
Comment (patrykmichalik) [inline vico/core/src/main/java/com/patrykandpatryk/vico/core/axis/horizontal/HorizontalAxis.kt:97]:
```suggestion
                .takeIf {
```

File: vico/core/src/main/java/com/patrykandpatryk/vico/core/axis/horizontal/HorizontalAxis.kt [hunk 3]
```diff
@@ -60,50 +72,66 @@ public class HorizontalAxis<Position : AxisPosition.Horizontal>(
...
                     top = chartBounds.top,
                     bottom = chartBounds.bottom,
                     centerX = tickCenter,
-                    boundingBox = chartBounds,
                 )
-            }?.drawVertical(context = context, top = chartBounds.top, bottom = chartBounds.bottom, centerX = tickCenter)
 
-            tick?.drawVertical(context = context, top = tickMarkTop, bottom = tickMarkBottom, centerX = tickCenter)
+            tick
+                ?.takeIf { shouldDraw }
```
Comment (patrykmichalik) [inline vico/core/src/main/java/com/patrykandpatryk/vico/core/axis/horizontal/HorizontalAxis.kt:114]:
```suggestion
                .takeIf { shouldDraw }
```

File: vico/core/src/main/java/com/patrykandpatryk/vico/core/axis/horizontal/HorizontalAxis.kt [hunk 4]
```diff
@@ -60,50 +72,66 @@ public class HorizontalAxis<Position : AxisPosition.Horizontal>(
...
+                ?.drawVertical(context = context, top = tickMarkTop, bottom = tickMarkBottom, centerX = tickCenter)
 
-            if (index < entryLength) {
-                label?.drawText(
+            label
+                ?.takeIf { index < entryLength && shouldDraw }
```
Comment (patrykmichalik) [inline vico/core/src/main/java/com/patrykandpatryk/vico/core/axis/horizontal/HorizontalAxis.kt:118]:
```suggestion
                .takeIf { index < entryLength && shouldDraw }
```

**Review by patrykmichalik — CHANGES_REQUESTED** [review]

File: vico/core/src/main/java/com/patrykandpatryk/vico/core/axis/horizontal/HorizontalAxis.kt (hunk 1 above)
Comment (Gowsky) [reply vico/core/src/main/java/com/patrykandpatryk/vico/core/axis/horizontal/HorizontalAxis.kt]: Done!

File: vico/core/src/main/java/com/patrykandpatryk/vico/core/axis/horizontal/HorizontalAxis.kt (hunk 2 above)
Comment (Gowsky) [reply vico/core/src/main/java/com/patrykandpatryk/vico/core/axis/horizontal/HorizontalAxis.kt:97]: `it` inside of `takeIf` lambda is required not to be null.

Comment (patrykmichalik) [reply vico/core/src/main/java/com/patrykandpatryk/vico/core/axis/horizontal/HorizontalAxis.kt:97]: Ah, sorry, my bad.

File: vico/compose/src/main/java/com/patrykandpatryk/vico/compose/axis/horizontal/HorizontalAxis.kt [hunk 5]
```diff
@@ -78,6 +82,8 @@ public fun topAxis(
  * @param axis the [LineComponent] to use for the axis line.
  * @param tick the [LineComponent] to use for ticks.
  * @param tickLength the length of ticks.
+ * @param tickPosition defines the position of ticks. [HorizontalAxis.TickPosition.Center] allows using custom offset
+ * and spacing for both ticks and labels.
```
Comment (patrykmichalik) [inline vico/compose/src/main/java/com/patrykandpatryk/vico/compose/axis/horizontal/HorizontalAxis.kt:86]:
```suggestion
 * @param tickPosition defines the position of ticks. [HorizontalAxis.TickPosition.Center] allows for using a custom offset
 * and spacing for both ticks and labels.
```

File: vico/compose/src/main/java/com/patrykandpatryk/vico/compose/axis/horizontal/HorizontalAxis.kt [hunk 6]
```diff
@@ -39,6 +39,8 @@ import com.patrykandpatryk.vico.core.component.text.TextComponent
...
```
Comment (patrykmichalik) [inline vico/compose/src/main/java/com/patrykandpatryk/vico/compose/axis/horizontal/HorizontalAxis.kt:43]:
```suggestion
 * @param tickPosition defines the position of ticks. [HorizontalAxis.TickPosition.Center] allows for using a custom offset
 * and spacing for both ticks and labels.
```

File: vico/core/src/main/java/com/patrykandpatryk/vico/core/axis/horizontal/HorizontalAxis.kt [hunk 7]
```diff
@@ -50,7 +52,17 @@ public class HorizontalAxis<Position : AxisPosition.Horizontal>(
     /**
      * Defines the tick placement.
      */
-    public var tickType: TickType = TickType.Minor
+    @Deprecated(message = "Tick type is now defined by `tickPosition`.", replaceWith = ReplaceWith("tickPosition"))
```
Comment (patrykmichalik) [inline vico/core/src/main/java/com/patrykandpatryk/vico/core/axis/horizontal/HorizontalAxis.kt:55]:
```suggestion
    @Deprecated(message = "The tick type is now defined by `tickPosition`.", replaceWith = ReplaceWith("tickPosition"))
```

File: vico/core/src/main/java/com/patrykandpatryk/vico/core/axis/horizontal/HorizontalAxis.kt [hunk 8]
```diff
@@ -242,9 +281,110 @@ public class HorizontalAxis<Position : AxisPosition.Horizontal>(
...
+                imports = arrayOf(
+                    "com.patrykandpatryk.vico.core.axis.horizontal.HorizontalAxis.TickPosition",
+                ),
+            ),
+        )
         Major,
     }
 
+    /**
+     * [TickPosition] defines the position of ticks. [HorizontalAxis.TickPosition.Center] allows for using custom offset
```
Comment (patrykmichalik) [inline vico/core/src/main/java/com/patrykandpatryk/vico/core/axis/horizontal/HorizontalAxis.kt:297]:
```suggestion
     * [TickPosition] defines the position of ticks. [HorizontalAxis.TickPosition.Center] allows for using a custom offset
```

File: vico/core/src/main/java/com/patrykandpatryk/vico/core/axis/horizontal/HorizontalAxis.kt [hunk 9]
```diff
@@ -242,9 +281,110 @@ public class HorizontalAxis<Position : AxisPosition.Horizontal>(
...
+     * @param spacing defines how often ticks should be drawn, where 1 means a tick is drawn for each entry,
+     * 2 means a tick is drawn for every second entry, and so on.
+     */
+    public sealed class TickPosition(
+        public val offset: Int,
+        public val spacing: Int,
+    ) {
+
+        /**
+         * Returns a tick count required by given [TickPosition].
```
Comment (patrykmichalik) [inline vico/core/src/main/java/com/patrykandpatryk/vico/core/axis/horizontal/HorizontalAxis.kt:310]:
```suggestion
         * Returns the tick count required by given [TickPosition].
```

File: vico/core/src/main/java/com/patrykandpatryk/vico/core/axis/horizontal/HorizontalAxis.kt [hunk 10]
```diff
@@ -242,9 +281,110 @@ public class HorizontalAxis<Position : AxisPosition.Horizontal>(
...
+         */
+        public abstract fun getTickCount(entryLength: Int): Int
+
+        /**
+         * Returns a chart bounds inset required by given [TickPosition].
```
Comment (patrykmichalik) [inline vico/core/src/main/java/com/patrykandpatryk/vico/core/axis/horizontal/HorizontalAxis.kt:315]:
```suggestion
         * Returns the chart bounds inset required by the given [TickPosition].
```

File: vico/core/src/main/java/com/patrykandpatryk/vico/core/axis/horizontal/HorizontalAxis.kt [hunk 11]
```diff
@@ -242,9 +281,110 @@ public class HorizontalAxis<Position : AxisPosition.Horizontal>(
...
+
+            override fun getTickCount(entryLength: Int): Int = entryLength
+
+            override fun getTickInset(tickThickness: Float): Float = 0f
+        }
+
+        public companion object {
+
+            /**
+             * Returns a [TickPosition] that replaces deprecated [type].
```
Comment (patrykmichalik) [inline vico/core/src/main/java/com/patrykandpatryk/vico/core/axis/horizontal/HorizontalAxis.kt:379]:
```suggestion
             * Returns the [TickPosition] that replaces the deprecated [type].
```

File: vico/core/src/main/java/com/patrykandpatryk/vico/core/axis/horizontal/HorizontalAxis.kt [hunk 12]
```diff
@@ -255,7 +395,13 @@ public class HorizontalAxis<Position : AxisPosition.Horizontal>(
         /**
          * Defines the tick placement.
          */
-        public var tickType: TickType = TickType.Minor
+        @Deprecated(message = "Tick type is now defined by `tickPosition`.", replaceWith = ReplaceWith("tickPosition"))
```
Comment (patrykmichalik) [inline vico/core/src/main/java/com/patrykandpatryk/vico/core/axis/horizontal/HorizontalAxis.kt:398]:
```suggestion
        @Deprecated(message = "The tick type is now defined by `tickPosition`.", replaceWith = ReplaceWith("tickPosition"))
```

**Review by patrykmichalik — CHANGES_REQUESTED** [review]

**Review by patrykmichalik — APPROVED** [review]

**Pull request merged**